[functorch] Prevent using for-loop for out-of-place index_fill batch rule

qqaatw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #99229

A follow-up PR for https://github.com/pytorch/pytorch/pull/91364#discussion_r1060723192


cc @zou3519 @Chillee @samdow @soumith @kshitij12345 @janeyx99

qqaatw committed:
[functorch] Prevent using for-loop for out-of-place index_fill batch rule

[ghstack-poisoned]

qqaatw committed:
Update on "[functorch] Prevent using for-loop for out-of-place index_fill batch rule"


A follow-up PR for https://github.com/pytorch/pytorch/pull/91364#discussion_r1060723192


[ghstack-poisoned]

qqaatw committed:
Update on "[functorch] Prevent using for-loop for out-of-place index_fill batch rule"


A follow-up PR for https://github.com/pytorch/pytorch/pull/91364#discussion_r1060723192


[ghstack-poisoned]

kshitij12345 commented on aten/src/ATen/functorch/BatchRulesScatterOps.cpp:
```diff
@@ -1123,20 +1124,48 @@ std::tuple<Tensor,optional<int64_t>> index_fill_int_tensor_batch_rule_impl(
+    return std::make_tuple(self_, 0);
+  }
 
   self_ = self_bdim.has_value() ? self_ : self_.clone();
```
IIUC, this line onwards is the same as `index_fill_int_scalar_batch_rule_impl` right? In that case, can we move this to helper function shared by both of them.
https://github.com/pytorch/pytorch/blob/beaaaaa9b4f8fdeddc21f5eef31fc3d786f3bad6/aten/src/ATen/functorch/BatchRulesScatterOps.cpp#L1067-L1093

kshitij12345 commented on aten/src/ATen/functorch/BatchRulesScatterOps.cpp:
```diff
@@ -1123,20 +1124,48 @@ std::tuple<Tensor,optional<int64_t>> index_fill_int_tensor_batch_rule_impl(
+      self_slice.index_fill_(
+        dim,
+        index_slice,
+        value_bdim.has_value() ? value_.select(0, i) : value_
```
Can we write this for-loop such that we won't have to check `value_bdim` every iteration?

qqaatw replied:
I feel this is the most concise version, and checking `value_bdim` would not be costly.

Here are the ways I came up:

1. Always ensure `value` has a batch dim and do select in every iteration - costly.
2. Make copies of the for-loop statement and separate the work out. There are four combinations for `inplace` and `value_bdim.has_value()` - not concise.
3. Determine `value_bdim.has_value()` before for-loop and use lambdas to wrap either `value_` or `value_.select(0, i)` - I feel unnecessary tbh. Would it be better?

Or maybe I'm missing something.

kshitij12345 replied:
Hmm, makes sense, plus I would assume that most of the time would be spent in `index_fill` so this should be good.

qqaatw committed:
Update on "[functorch] Prevent using for-loop for out-of-place index_fill batch rule"


A follow-up PR for https://github.com/pytorch/pytorch/pull/91364#discussion_r1060723192


cc zou3519 Chillee samdow soumith kshitij12345 janeyx99

[ghstack-poisoned]

qqaatw committed:
Update on "[functorch] Prevent using for-loop for out-of-place index_fill batch rule"


A follow-up PR for https://github.com/pytorch/pytorch/pull/91364#discussion_r1060723192


cc zou3519 Chillee samdow soumith kshitij12345 janeyx99

[ghstack-poisoned]

kshitij12345 commented on aten/src/ATen/functorch/BatchRulesScatterOps.cpp:
```diff
@@ -1123,22 +1136,27 @@ std::tuple<Tensor,optional<int64_t>> index_fill_int_tensor_batch_rule_impl(
-  return std::make_tuple(self_, 0);
+  self_ = self_bdim.has_value() ? self_ : self_.clone();
+
+  return index_fill_batch_rule_helper(batch_size, self_logical_rank, index_logical_rank, self_, dim, index_, value.item());
```
We should add a comment here that calling item on value is fine because `value` is guaranteed to be a regular tensor and not a BatchedTensor (because `item` does not work with BatchedTensor)

Thank you!

qqaatw replied:
done.

qqaatw committed:
Update on "[functorch] Prevent using for-loop for out-of-place index_fill batch rule"


A follow-up PR for https://github.com/pytorch/pytorch/pull/91364#discussion_r1060723192


cc zou3519 Chillee samdow soumith kshitij12345 janeyx99

[ghstack-poisoned]

kshitij12345 reviewed (approved):
LGTM! Thank you @qqaatw

kshitij12345 commented:
@pytorchbot merge
